Fix powsimp issues #9183, #10095, and KeyError

ylemkimon opened the pull request:
- Fixed powsimp raising ValueError when base is float and Pow is argument of Mul. Since multiplicity expects integer or rational but powsimp doesn't checks it, it raises ValueError: fixes #9183
- Fixed Pow.as_numer_denom() returning nan when either numerator or denominator of base is 1 and exp is infinite. Since 1**oo is NaN, numerator or denominator becomes NaN: fixes #10095
- Fixed powsimp raising KeyError in common_b[b] = common_b[b] + e, when one base is Float and another is equal Rational(Integer). This seems to be caused by #11707.

~**Thing(s) to consider:** Should we automatically evaluate k**oo to zero where k is in (-1, 1), i.e., `(1/(2*E))**oo`?~

ylemkimon committed:
Fix Pow.as_numer_denom() when exp is infinite

Fixed Pow.as_numer_denom() returning nan when either numerator or
denominator is 1.

ylemkimon committed:
Fix powsimp raising ValueError when base is float

Fixed powsimp raising ValueError: expecting ints or fractions ~ when
base is float.

ylemkimon committed:
Add tests for #9183 and #10095

Added tests for #9183 and #10095 fixed by 'Fix Pow.as_numer_denom() when
exp is infinite' and 'Fix powsimp raising ValueError when base is
float'.

ylemkimon committed:
Fix test case for #10095

Fixed test case for #10095.

asmeurer commented:
> Should we automatically evaluate k**oo to zero where k is in (-1, 1), 

I don't think automatic evaluation should use evalf at all. That can be slow in general and automatic stuff should be fast (see also https://github.com/sympy/sympy/wiki/Automatic-Simplification). Those examples you gave all work because the power can be implemented directly on the Number object itself without any actual numeric evaluation (for instance, `E` (actually `exp`) knows that the power of `-oo` is 0.

Inside a simplification function like powsimp however it makes sense.

ylemkimon committed:
Fix powsimp raising KeyError when base has Float

Fixed powsimp raising KeyError in common_b[b] = common_b[b] + e, when
one base has Float and another has its Rational(Integer) counterpart.
This seems to be caused by sympy/sympy#11707.

ylemkimon commented:
@asmeurer I've been mistaken assumptions like .is_positive do not use evalf. I'll create new issue regarding it.

---
Added another commit: Fixed powsimp raising KeyError in common_b[b] = common_b[b] + e, when one base has Float and another has its Rational(Integer) counterpart. This seems to be caused by #11707.

ylemkimon commented:
@asmeurer Please review. Thank you.

asmeurer commented on sympy/simplify/powsimp.py:
```diff
@@ -163,9 +163,9 @@ def recurse(arg, **kwargs):
-            if (b and b.is_Number and not all(ei.is_Number for ei in e) and \
-                    coeff is not S.One and
-                    b not in (S.One, S.NegativeOne)):
+            if (b and b.is_Number and b.is_rational and not all(
```
If you need an actual rational number you should use is_Rational or isinstance(b, Rational). is_rational will also be True for things like `Symbol('x', rational=True)`.

ylemkimon replied:
@asmeurer Fixed 😄

asmeurer commented:
I have just the one comment. +1 to merge once it is fixed.

ylemkimon committed:
Change is_Number and is_rational to is_Rational

Changed is_Number and is_rational to is_Rational

ylemkimon commented:
@asmeurer Fixed 😄
